File: csharp/HomeScheduler/HomeScheduler.Data/Repository/UserRepository.cs
using HomeScheduler.Data.Entity;
using Microsoft.EntityFrameworkCore.ChangeTracking;
using Microsoft.Extensions.DependencyInjection;

namespace HomeScheduler.Data.Repository;

public class UserRepository : IUserRepository
{
    private readonly IServiceScopeFactory _serviceScopeFactory;

    public UserRepository(IServiceScopeFactory serviceScopeFactory)
    {
        _serviceScopeFactory = serviceScopeFactory;
    }

    public List<User> GetUsers()
    {
        using IServiceScope scope = _serviceScopeFactory.CreateScope();
        using HomeSchedulerContext context = scope.ServiceProvider.GetRequiredService<HomeSchedulerContext>();

        return context.Users.ToList();
    }

    public User? GetUser(int id)
    {
        using IServiceScope scope = _serviceScopeFactory.CreateScope();
        using HomeSchedulerContext context = scope.ServiceProvider.GetRequiredService<HomeSchedulerContext>();
   
        return context.Users.FirstOrDefault(x => x.UserId == id);
    }

    public UpdateResponse<User> AddUser(User entity)
    {
        using IServiceScope scope = _serviceScopeFactory.CreateScope();
        using HomeSchedulerContext context = scope.ServiceProvider.GetRequiredService<HomeSchedulerContext>();

        EntityEntry<User> entry = context.Users.Add(entity);
        int changeCount = context.SaveChanges();

        return new UpdateResponse<User>
        {
            Data = entry.Entity,
            ChangeCount = changeCount
        };
    }

    public UpdateResponse<User> UpdateUser(int id, User entity)
    {
        using IServiceScope scope = _serviceScopeFactory.CreateScope();
        using HomeSchedulerContext context = scope.ServiceProvider.GetRequiredService<HomeSchedulerContext>();

        User? existingEntity = context.Users.FirstOrDefault(x => x.UserId == id);

        if (existingEntity == null)
        {
            return new UpdateResponse<User>();
        }

        existingEntity.UserName = entity.UserName;
        existingEntity.FirstName = entity.FirstName;
        existingEntity.LastName = entity.LastName;
        existingEntity.UpdatedBy = entity.UpdatedBy;
        existingEntity.UpdatedDate = entity.UpdatedDate;

        EntityEntry<User> entry = context.Users.Update(existingEntity);
        int changeCount = context.SaveChanges();

        return new UpdateResponse<User>
        {
            Data = entry.Entity,
            ChangeCount = changeCount
        };
    }
}

File: python-django/api/data/UserRepository.py
from ..models import User

from .UpdateResponse import UpdateResponse


class UserRepository:
    def get_users(self):
        return User.objects.all()

    def get_user(self, user_id):
        return User.objects.get(pk=user_id)

    def add_user(self, entity: User) -> UpdateResponse[User]:
        entity.save()
        return UpdateResponse(entity, 1)

    def update_user(self, user_id, entity: User) -> UpdateResponse[User]:
        existing_entity = User.objects.get(pk=user_id)
        if (existing_entity is None):
            return UpdateResponse(None, 0)

        existing_entity.username = entity.username
        existing_entity.firstname = entity.firstname
        existing_entity.lastname = entity.lastname
        existing_entity.updatedby = entity.updatedby
        existing_entity.updateddate = entity.updateddate

        existing_entity.save()
        return UpdateResponse(existing_entity, 1)
